cpp/src/arrow/compute/kernels/comparator.cc

```diff
@@ -0,0 +1,197 @@
+#include "arrow/util/bit-util.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+static inline Status CopyOrTransferBitmap(const Array& array, MemoryPool* pool,
+                                          std::shared_ptr<Buffer>* bitmap) {
+  auto offset = array.offset();
+  auto length = array.length();
+  auto null_bitmap = array.null_bitmap();
+
+  if (null_bitmap == nullptr || offset == 0 || length == 0) {
+    *bitmap = null_bitmap;
+    return Status::OK();
+  }
+
+  return internal::CopyBitmap(pool, null_bitmap->data(), offset, length, bitmap);
+}
+
+template <typename ArrowType, CompareOperator Operator,
+          typename ArrayType = typename TypeTraits<ArrowType>::ArrayType,
+          typename ScalarType = typename TypeTraits<ArrowType>::ScalarType,
+          typename T = typename TypeTraits<ArrowType>::CType>
+static Status CompareArrayScalar(const ArrayType& input, const ScalarType& scalar,
+                                 uint8_t* bitmap) {
+  const T right = scalar.value;
+  const T* values = input.raw_values();
+
+  // Bitmap is not sliced like the original array, thus offset=0
+  auto writer = internal::BitmapWriter(bitmap, 0, input.length());
+  for (int64_t i = 0; i < input.length(); i++, writer.Next()) {
+    if (Comparator<T, Operator>::Compare(values[i], right)) {
+      writer.Set();
+    } else {
+      writer.Clear();
+    }
+  }
+
+  writer.Finish();
+
+  return Status::OK();
+}
+
+template <typename ArrowType, CompareOperator Op>
+class CompareFunction final : public FilterFunction {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
+
+ public:
+  explicit CompareFunction(FunctionContext* ctx) : ctx_(ctx) {}
+
+  Status Filter(const Array& input, const Scalar& scalar, ArrayData* output) const {
+    // Caller must cast
+    DCHECK(input.type()->Equals(scalar.type));
+    // Caller must not pass null scalar, use IsNull, IsNotNull
+    DCHECK(scalar.is_valid);
+    // Output must be a boolean array
+    DCHECK(output->type->Equals(boolean()));
+    // Output must be of same length
+    DCHECK_EQ(output->length, input.length());
+
+    // Copy null_bitmap
+    RETURN_NOT_OK(CopyOrTransferBitmap(input, ctx_->memory_pool(), &output->buffers[0]));
+
+    DCHECK(output->buffers[1]);
+    uint8_t* bitmap_result = output->buffers[1]->mutable_data();
+    return CompareArrayScalar<ArrowType, Op>(static_cast<const ArrayType&>(input),
+                                             static_cast<const ScalarType&>(scalar),
+                                             bitmap_result);
+  }
+
+ private:
+  FunctionContext* ctx_;
+};
+
+template <typename ArrowType, CompareOperator Op>
+static inline std::shared_ptr<FilterFunction> MakeCompareFunctionTypeOp(
+    FunctionContext* ctx) {
+  return std::static_pointer_cast<FilterFunction>(
```

Comment:
This cast is unnecessary- `shared_ptr` (and `unique_ptr`) support implicit upcasting